python/lsst/daf/butler/core/datasets/type.py

```diff
@@ -302,33 +291,28 @@ def isCalibration(self) -> bool:
         """
         return self._isCalibration
 
-    def finalizeParentStorageClass(self, newParent: StorageClass) -> None:
-        """Finalize the parent storage class definition.
+    def replaceParentStorageClass(self, newParent: StorageClass) -> None:
+        """Replace the parent storage class definition.
 
-        Replaces the current placeholder parent storage class with
-        the real parent.
+        Replaces the current parent storage class with a new parent, this
+        could cause failures when interacting with the registry. Only use
+        this method if you know what you are doing.
```

Comment:
This really scares me; `DatasetType` is otherwise immutable.  I'm guessing it's probably okay because this is only used on refs produced by `GraphBuilder` before they are ever released to the outside world, but there are still a lot of moving parts there that could get confused if something changes under them.

If we can't do this without breaking immutability, I think I might be happier if we just let `GraphBuilder` poke around at private `DatasetType` attributes directly so it's clear it's being evil, instead of sanitizing that evil method (even one with a prominent warning).